File: mobile/app/src/main/java/br/edu/fatecpg/estoque_inteligente/utils/ColorsProvider.kt
package br.edu.fatecpg.estoque_inteligente.utils

import br.edu.fatecpg.estoque_inteligente.model.Color

class ColorsProvider {
    fun stringToColor(input: String): Color {
        val hash = input.hashCode()
        val r = (hash shr 16) and 0xFF
        val g = (hash shr 8) and 0xFF
        val b = hash and 0xFF
        return Color(r, g, b)
    }

    fun getContrastColor(bg: Color): Color {
        val luminance = 0.299 * bg.red + 0.587 * bg.green + 0.114 * bg.blue
        return if (luminance > 128) Color(0, 0, 0) else Color(255, 255, 255)
    }

    fun pastelize(c: Color, factor: Float = 0.7f): Color {
        // factor: how much white to mix in, 0 = no change, 1 = full white
        val r = (c.red * (1 - factor) + 255 * factor).toInt().coerceIn(0, 255)
        val g = (c.green * (1 - factor) + 255 * factor).toInt().coerceIn(0, 255)
        val b = (c.blue * (1 - factor) + 255 * factor).toInt().coerceIn(0, 255)
        return Color(r, g, b)
    }

    fun stringToThemeColors(input: String): Pair<Color, Color> {
        val background = pastelize(stringToColor(input))
        val foreground = getContrastColor(background)
        return Pair(background, foreground)
    }
}

File: mobile-react/src/utils/colorsProvider.ts
export type RGB = { red: number; green: number; blue: number };

export const stringToColor = (input: string): RGB => {
  let hash = 0;
  for (let i = 0; i < input.length; i++) {
    hash = input.charCodeAt(i) + ((hash << 5) - hash);
  }
  const r = (hash >> 16) & 0xff;
  const g = (hash >> 8) & 0xff;
  const b = hash & 0xff;
  return { red: r, green: g, blue: b };
};

export const getContrastColor = (bg: RGB): RGB => {
  const luminance = 0.299 * bg.red + 0.587 * bg.green + 0.114 * bg.blue;
  return luminance > 128 ? { red: 0, green: 0, blue: 0 } : { red: 255, green: 255, blue: 255 };
};

export const pastelize = (c: RGB, factor = 0.7): RGB => {
  const r = Math.round(c.red * (1 - factor) + 255 * factor);
  const g = Math.round(c.green * (1 - factor) + 255 * factor);
  const b = Math.round(c.blue * (1 - factor) + 255 * factor);
  return { red: Math.min(255, Math.max(0, r)), green: Math.min(255, Math.max(0, g)), blue: Math.min(255, Math.max(0, b)) };
};

export const stringToThemeColors = (input: string) => {
  const bg = pastelize(stringToColor(input));
  const fg = getContrastColor(bg);
  return { bg, fg };
};
